common_server/src/main/java/io/pravega/common/util/BufferViewComparator.java

```diff
@@ -0,0 +1,320 @@
+ * - Define A{n} and B{n} as the bit at position n in A and B, respectively. This can be either 0 or 1.
+ * - A precedes B if there exists bit position i such that for all bit positions j smaller than i,
+ * then A{j} is equal to B{j}, A{i} is 0 and B{i} is 1.
+ * - A is equal to B if the values of all bit positions in both {@link BufferView}s match.
+ * <p>
+ * Lexicographic bitwise comparison for {@link BufferView}s of different lengths:
+ * - Consider two {@link BufferView}s A and B, with A having LA bits and B having LB bits (LA, LB are multiples of 8).
+ * - We do a Lexicographic bitwise comparison of the prefixes of A and B of lengths Min(LA, LB).
+ * - If the prefixes are equal, then the shorter of A and B precedes the longer of A and B.
+ * -- If LA &lt; LB, then A is before B; if LA &gt; LB, then A is after B.
+ * - If the prefixes are not equal, then the result from the prefix comparison is used to order A and B (see above).
+ * <p>
+ * Lexicographic bitwise comparison matches the natural order of numbers when serialized as unsigned (i.e., using the
+ * specialized methods in {@link BitConverter}) since they avoid the complications involved with interpreting individual
+ * bytes with the first bit set to 1 using 2's complement (128 is before 127 if we used signed bytes).
+ * <p>
+ * For example:
+ * - Consider any two Longs L1 and L2.
+ * - Let S1 be the result of {@link BitConverter#writeUnsignedLong} when applied to L1, and S2 the result when applied to L2.
+ * - Then {@link Long#compare} applied to (L1, L2) is equal to {@link #compare} applied to (S1, S2).
+ * - This equality would not hold should L1 and L2 be serialized using {@link BitConverter#writeLong} or if we used plain
+ * (signed) byte comparison internally.
+ */
+public abstract class BufferViewComparator implements Comparator<byte[]>, Serializable {
+    /**
+     * The minimum byte value for this comparison. Since we use unsigned bytes, this is 0-based.
+     */
+    public static final byte MIN_VALUE = 0;
+    /**
+     * The maximum byte value for this comparison. Since we use unsigned bytes, this is 0-based, hence 255.
+     * Note that the actual value stored in Java for this is 0xFF, which is actually -1.
+     */
+    public static final byte MAX_VALUE = (byte) 255;
+
+    private static final long serialVersionUID = 1L;
+
+    //region Constructor
+
+    /**
+     * Determines if Intrinsic (JRE9+) array and {@link ByteBuffer} comparison is available.
+     */
+    private static final boolean INTRINSIC_SUPPORTED;
+
+    static {
+        // Figure out if we support JRE9+ Intrinsic comparators. We do compile this code with JDK11, however this is to
+        // have a failover in case we run with a lesser JRE.
+        boolean intrinsicSupported = true;
+        try {
+            ByteBuffer bb = ByteBuffer.wrap(new byte[1]);
+            bb.mismatch(bb);
+        } catch (ExceptionInInitializerError | UnsatisfiedLinkError | NoClassDefFoundError e) {
```

Comment:
What about using multirelease jars? It will be more simpler at runtime and less error prone

Reply:
Gradle [recommends against that](https://blog.gradle.org/mrjars), so in my opinion, this is less controversial (and more deterministic - what goes where).